docs/e2e-testing.md

```diff
@@ -0,0 +1,208 @@
+- Adopt a fail-fast philosophy: If an initial step in a test sequence fails, it may not be logical or efficient to proceed with subsequent steps. In such cases, adopting test coupling can be beneficial. However, consider the impact of a failure. If a failure in one part of a combined test makes it impossible to test the rest, but you still need to test the subsequent parts regardless of the outcome of the first part, it might be better to isolate the tests.
+
+Remember, the goal is to create tests that are reliable, easy to understand, and provide valuable feedback about the system. Whether you choose to isolate or combine tests will depend on what you're trying to achieve within your tests. Ideally, you should aim for a large number of unit tests to test individual code pieces and a medium number of E2E tests for user flow testing that cover all the scenarios.
+
+## Controlling state
+
+To achieve test atomicity and ensure E2E tests are stable and reliable, it is best to control the state of the application programmatically rather than relying on the application UI. Doing so eliminates unnecessary UI interactions, decreasing the amount of possible breaking points in a test. It also improves test stability by minimising issues caused by timing synchronisation or inconsistencies in the UI, reducing the test execution time and allowing the test to provide fast and focused feedback.
+
+### Guidelines
+
+✅ Use fixture to remove multiple redundant steps: Add Contact, Open test dapp, Connect to test dapp, Deploy TST, Add TST to wallet
+
+```javascript
+new FixtureBuilder()
+  .withAddressBookControllerContactBob()
+  .withTokensControllerERC20()
+  .build();
+```
+
+Send TST test steps after setting this fixture:
+
+```javascript
+Login
+Send TST
+Assertion
+```
+
+⚠️ Use fixture to remove single redundant steps: Add Contact
+
+```javascript
+new FixtureBuilder().withAddressBookControllerContactBob().build();
+```
+
+Send TST test steps after setting this fixture:
+
+```javascript
+Login
+Open test dapp
+Connect to test dapp
+Deploy TST
+Add TST to wallet
+Send TST
+Assertion
+```
+
+❌ Use the UI to build state
+
+```javascript
+new FixtureBuilder().build();
+```
+
+Send TST test steps after setting this fixture:
+
+```javascript
+Login
+Add Contact
+Open test dapp
+Connect to test dapp
+Deploy TST
+Add TST to wallet
+Send TST
+Assertion
+```
+
+Here are some examples to remove redundant steps following above guidelines:
+
+1.
+
+```javascript
+// test file: test/e2e/tests/multiple-transactions.spec.js
+scenario: creates multiple queued transactions and then confirms.
+solution: Ensure tests are different from the tests that line in test/e2e/tests/navigate-transactions.spec.js
+```
+
+2.
+
+```javascript
+// test file: test/e2e/tests/multiple-transactions.spec.js
```

Comment:
```suggestion
// test file: metamask-extension/test/e2e/tests/multiple-transactions.spec.js
```

Reply:
As discussed in our 1<>1, we opted to omit the root folder in the relative path and specify the path of the e2e tests. I have updated this.